homeassistant/components/ws66i/media_player.py

```diff
@@ -0,0 +1,213 @@
+"""Support for interfacing with WS66i 6 zone home audio controller."""
+from copy import deepcopy
+import logging
+
+from pyws66i import WS66i, ZoneStatus
+
+from homeassistant.components.media_player import MediaPlayerEntity
+from homeassistant.components.media_player.const import (
+    SUPPORT_SELECT_SOURCE,
+    SUPPORT_TURN_OFF,
+    SUPPORT_TURN_ON,
+    SUPPORT_VOLUME_MUTE,
+    SUPPORT_VOLUME_SET,
+    SUPPORT_VOLUME_STEP,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import STATE_OFF, STATE_ON
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers.entity import DeviceInfo
+from homeassistant.helpers.entity_platform import (
+    AddEntitiesCallback,
+    async_get_current_platform,
+)
+from homeassistant.helpers.update_coordinator import CoordinatorEntity
+
+from .const import DOMAIN, SERVICE_RESTORE, SERVICE_SNAPSHOT
+from .coordinator import Ws66iDataUpdateCoordinator
+from .models import Ws66iData
+
+_LOGGER = logging.getLogger(__name__)
+
+PARALLEL_UPDATES = 1
+
+SUPPORT_WS66I = (
+    SUPPORT_VOLUME_MUTE
+    | SUPPORT_VOLUME_SET
+    | SUPPORT_VOLUME_STEP
+    | SUPPORT_TURN_ON
+    | SUPPORT_TURN_OFF
+    | SUPPORT_SELECT_SOURCE
+)
+
+MAX_VOL = 38
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the WS66i 6-zone amplifier platform from a config entry."""
+    ws66i_data: Ws66iData = hass.data[DOMAIN][config_entry.entry_id]
+
+    # Build and add the entities from the data class
+    async_add_entities(
+        Ws66iZone(
+            device=ws66i_data.device,
+            ws66i_data=ws66i_data,
+            entry_id=config_entry.entry_id,
+            zone_id=zone_id,
+            data_idx=idx,
+            coordinator=ws66i_data.coordinator,
+        )
+        for idx, zone_id in enumerate(ws66i_data.zones)
+    )
+
+    # Set up services
+    platform = async_get_current_platform()
+
+    platform.async_register_entity_service(
+        SERVICE_SNAPSHOT,
+        {},
+        "snapshot",
+    )
+
+    platform.async_register_entity_service(
+        SERVICE_RESTORE,
+        {},
+        "async_restore",
+    )
```

Comment:
I have these two services because I modeled this integration after the monoprice integration.

I was just playing with scenes after you mentioned this. I couldn't see a way for the scene to save the state of the entity at the time of an automation being triggered. A use case of the snapshots/restore is:

1. User is listening to background music in the kitchen on source 3, volume 10 (arbitrary values)
2. Doorbell is rung
3. An automation is triggered
   a. The snapshot service is called to save the state
   b. A scene is loaded, which sets the source to 4, volume to 20
   c. A media file is played ("ding-dong" notification sound)
   d. The restore service is called to revert state back to source 3, volume 10

The user is notified that someone is at the door, and the background music resumes. Can this behavior be replicated without these services?

Reply:
That is exactly the services scene provides and was created for. Scenes allow you to capture states and restore states (it actually uses state restore under the hood).

For example, you could create a snapshot of a speaker, before doing anything:

```yaml
- service: scene.create
  data:
    scene_id: snapshot_of_sonos_office
    snapshot_entities:
      - media_player.sonos_office
```

Next, do the stuff one wants... and after that sequence is done, restore the previous snapped scene:

```yaml
- service: scene.turn_on
  target:
    entity_id: scene.snapshot_of_sonos_office
```

Reply:
Perfect! Thanks for the thorough response. Those custom services are not needed. I'll remove them and add that change to the PR that @MartinHjelmare is requesting.